### ISSUE TITLE
Pause plot programmatically

### ISSUE THREAD

**Issue by aarondfrancis** [evil-mad/axidraw#75]
Hey there,

Is there a way to pause a plot via GPIO or the CLI as if the physical button had just been pressed? 

I see that there is some sort of "[alternate pause](https://www.schmalzhaus.com/EBB/EBBCommands.html)" available

> SC,13,<use_alt_pause> - turns on (1) or off (0) alternate pause button function on RB0. On by default. For v1.1 boards, it uses RB2 instead. (new in v2.0)

but am not quite sure how to use it. I'm not sure where the RB2 pin is either.

Thanks in advance for any help!

**Comment by oskay** [comment]
Can you please clarify your question?  What is it that you are trying to pause, and how do you want to pause it?  Are you plotting from Inkscape, or from the CLI, or "via GPIO"?

Also: The document that you have linked is (as it says there) deprecated in its entirety. Please refer to the current EBB documentation, hosted at: http://evil-mad.github.io/EggBot/ebb.html

**Comment by aarondfrancis** [comment]
Sure... I can try.

When a file is being plotted, one can push the pause button and it will pause and store the appropriate resume data. 

Is there a way to pause a currently running plot _without_ pushing the button? (And not by changing the layer name to have a preceding `!`.)

I'm happy with any method you can provide. If it's possible using the CLI I can do that, if it's possible by using some combo of GPIO and the RB0 (?) pin, I can do that.

I've read that there is an alternate pause functionality, from the documentation you linked: 

> SC,13,use_alt_pause - turns on (1) or off (0) alternate pause button function on RB0. On by default. For EBB v1.1 boards, it uses RB2 instead.

I'm not sure how to use that though. Since it's on by default, if I hit the RB0 pin (wherever that is) with a HIGH, will the axidraw pause itself (and store the resume data) as if the physical button was pressed?

Hope that's clearer.

Thanks

**Comment by oskay** [comment]
Are you plotting from the AxiDraw CLI or from Inkscape, or something else?

**Comment by aarondfrancis** [comment]
CLI

**Comment by oskay** [comment]
And you would like to provide an externally-generated I/O signal to control initiate a pause?

**Comment by aarondfrancis** [comment]
I can if that's the only way! Ideally I'd like to signal the running process itself with e.g. SIGINT, as that'd be easier. But if that's not possible an I/O signal to the board is fine.

**Comment by oskay** [comment]
[SIGINT](https://en.wikipedia.org/wiki/Signals_intelligence)?

There is not presently any mechanism in place for a third-party process to gracefully stop (or otherwise modify) a plot in progress via the CLI. However, the code is open source, so you might want to consider adding a check for a separate internal signal in parallel with the "button press" detection that works with your application.

The "alternate pause" feature on B0 works as follows: There is an internal pull-up on input pin B0. If you ground that pin, then the EBB will record that event as though you had pressed the PRG button. (We apparently need to document this better in the EBB command set documentation.)

The set of three pins directly above the three pins used for the servo connector are the "B0" connector. These three pins are ground, +5V, signal B0, reading from the edge of the board inward.

Here is an example photo showing an external pause button connected to B0:
![IMG_1699](https://user-images.githubusercontent.com/1009180/73974202-cac0f300-48d8-11ea-9465-5aaaaf8aab31.jpg)

**Comment by aarondfrancis** [comment]
😂 

Haha way less interesting than that, SIGINT like `kill -2` https://pubs.opengroup.org/onlinepubs/7908799/xcu/kill.html.

My ideal would be starting a process via CLI and then sending an interrupt signal would pause the process, allowing the axidraw to be sent back to home position. Currently if I need to stop a plot I send `kill -9` which means the internal positioning is lost and I have to move it back home manually.

I'll take a look into the CLI code and see if I can find where the button press code is and perhaps weasel a solution in there. I don't know python very well, but it can't be too difficult to read it.

Thanks for the info on the B0 pin, that's really helpful. I'm running the whole thing off a Pi so I may look at wiring one of the Pi's GPIO pins to the B0 and then I can gracefully pause it that way. 

Those are the things I needed to know, I appreciate the help!

**Comment by oskay** [comment]
Great. We'll look into adding a software-controlled pause in the future. The hardest part from my perspective is to see *how* it would be initiated-- what the interface would look like.

**Comment by aarondfrancis** [comment]
https://stackoverflow.com/a/1112350/1408651

This is how I envision it:

- User starts a plot with `python axicli.py file.svg`
- Axidraw starts
- User hits Ctrl+C for some reason
- CLI behaves exactly as if the physical button on the EBB was pressed.

**Comment by aarondfrancis** [comment]
Last question!

I'm trying to add an option `--pause_if_exists` to the CLI, but I can't seem to get it to pass all the way through for the life of me.

I've tried this in  `axidraw.py`

```
self.OptionParser.add_option("--pause_if_exists",\
                    type="string", action="store", dest="pause_if_exists",\
                    help="Pause the plot if this file exists")
```

and this in `core.py`

```
parser.add_argument("--pause_if_exists",\
                help="Pause if file exists")
```

and I see there is also `common_options.py`?

Ultimately, I'm trying to add 

```
if self.options.pause_if_exists is not None and path.exists(self.options.pause_if_exists):
            pause_state = 1
```

in `axidraw.py`.  I've gotten the axidraw to pause by hardcoding the file path in the CLI, but now I'm trying to pass a path in and I can't seem to decipher where all I need to add the option code.

Thanks!

**Comment by oskay** [comment]
Any reason not to just use two files there, one to enable the option, the other to assert pause?

**Comment by aarondfrancis** [comment]
I'm controlling multiple units from the Pi so I need them to be
individually paused.

I tried using the named port to generate the pause file filename but it's
nulled out by that point

**Comment by oskay** [comment]
There are a few layers deep that you need to feed those options, including the CLI core, axidraw_control, and axidraw. 

As an alternative, I think that you *might* be able to use different configuration files to specify this. If you use a different config file per machine -- identical except for port name -- that should work, and I expect that axidraw_conf.port is not nulled out.


We've often seen driving one AxiDraw per Pi, but it's new for us seeing multiple AxiDraws per Pi.

**Comment by aarondfrancis** [comment]
Interesting. I will try that tomorrow morning. Thanks!

**Comment by aarondfrancis** [comment]
Solved it!

I thought that adding a param would be the easiest option, but as you said there are so many layers to get that to work.

It turns out the SIGINT was trivial. Here's what I ended up adding (along with a few imports at the top)

![image](https://user-images.githubusercontent.com/881931/74051907-8009ae80-499e-11ea-9cc1-a49eafdd3e2c.png)

Not sure if this is PR worthy, as I'm unsure how common the use-case is. Regardless, it's working for me now! 

I can send a `kill -2 [pid]` and it exits gracefully.

Thanks for all your help!
